Mathlib/Algebra/Lie/LieTheorem.lean

```diff
@@ -0,0 +1,554 @@
+
+lemma trace_T_res_zero (hv : v ∈ altWeightSpace A χ) :
+    LinearMap.trace k (iSupU v (π k V z)) ((T A χ w).restrict (T_map_iSupU A χ z w hv)) = 0 := by
+  apply IsNilpotent.eq_zero
+  exact LinearMap.isNilpotent_trace_of_isNilpotent (T_res_nilpotent A χ z w hv)
+
+lemma πaz_map_iSupU (a : A) (hv : v ∈ altWeightSpace A χ) :
+    ∀ x ∈ iSupU v (π k V z), π k V ⁅a,z⁆ x ∈ iSupU v (π k V z):= by
+  intro x hx
+  apply iSupU_A_stable A χ z ⟨⁅a,z⁆, lie_mem_left k L A a z a.prop⟩ hv
+  assumption
+
+lemma trace_πaz (a : A) (hv : v ∈ altWeightSpace A χ):
+    LinearMap.trace k (iSupU v (π k V z)) ((π k V ⁅a, z⁆).restrict (πaz_map_iSupU A χ z a hv))
+    = χ ⟨⁅a,z⁆, lie_mem_left k L A a z a.prop⟩ • (finrank k (iSupU v (π k V z))) := by
+  rw [← LinearMap.trace_id, ← LinearMap.map_smul, ← sub_eq_zero, ← LinearMap.map_sub]
+  apply trace_T_res_zero A χ z ⟨⁅a,z⁆, lie_mem_left k L A a z a.prop⟩ hv
+
+theorem trace_πaz_zero (a : A) (hv : v ∈ altWeightSpace A χ):
+    LinearMap.trace k (iSupU v (π k V z)) ((π k V ⁅a, z⁆).restrict (πaz_map_iSupU A χ z a hv))
+    = 0 := by
+  have hzU : ∀ x ∈ iSupU v (π k V z), (π k V z) x ∈ iSupU v (π k V z) :=
+    fun _ hx ↦ (map_iSupU_le_iSupU v (π k V z)) (Submodule.mem_map_of_mem hx)
+  have hres : (π k V ⁅a, z⁆).restrict (πaz_map_iSupU A χ z a hv) =
+    ⁅(π k V a).restrict (iSupU_A_stable A χ z a hv),
+    (π k V z).restrict hzU⁆ := by
+    ext ⟨x, hx⟩
+    simp
+  rw [hres, LieRing.of_associative_ring_bracket, map_sub, LinearMap.trace_mul_comm, sub_self]
+
+lemma chi_az_zero (a : A) (hv : v ∈ altWeightSpace A χ) (hv' : v ≠ 0):
+    χ ⟨⁅a,z⁆, lie_mem_left k L A a z a.prop⟩ = 0 := by
+  have h := trace_πaz A χ z a hv
+  rw [trace_πaz_zero A χ z a hv] at h
+  suffices h' : finrank k ↥(iSupU v (π k V z)) ≠ 0 by
+    aesop
+  have hvU : v ∈ iSupU v (π k V z) := by
+    apply Submodule.mem_iSup_of_mem 1
+    apply Submodule.subset_span
+    use 0, zero_lt_one
+    rw [pow_zero, LinearMap.one_apply]
+  have iSupU_nontrivial : Nontrivial (iSupU v (π k V z)) := ⟨⟨v,hvU⟩,0, ?_⟩
+  swap
+  simp only [ne_eq, Submodule.mk_eq_zero, hv', not_false_eq_true]
+  apply Nat.ne_of_lt'
+  apply FiniteDimensional.finrank_pos
+
+theorem altWeightSpace_lie_stable (hv : v ∈ altWeightSpace A χ):  ⁅z, v⁆ ∈ altWeightSpace A χ := by
+  rcases eq_or_ne v 0 with (rfl | hv')
+  · simp only [lie_zero, Submodule.zero_mem]
+  · intro a
+    have hzwv : ⁅⁅a.val, z⁆, v⁆ = χ ⟨⁅a,z⁆, lie_mem_left k L A a z a.prop⟩ • v :=
+      hv ⟨⁅a,z⁆, lie_mem_left k L A a z a.prop⟩
+    rw [leibniz_lie, hv a, hzwv, chi_az_zero A χ z a hv hv']
+    simp only [zero_smul, lie_smul, zero_add]
+end
+
+instance : IsCoatomic (Submodule k V) :=
+  isCoatomic_of_isAtomic_of_complementedLattice_of_isModular
+
+lemma exists_coatom (n : ℕ) (hV : finrank k V = n + 1) (X : Submodule k V) (hX : X ≠ ⊤) :
+    ∃ W : Submodule k V, X ≤ W ∧ finrank k W = n ∧ IsCoatom W := by
+  obtain (h1 | ⟨W, hW1, hW2⟩) := eq_top_or_exists_le_coatom X; contradiction
+  use W, hW2
+  have hW1' := hW1
```

Comment:
I create a copy of `hW1` so that I can transform one of them with rewrites and keep the other as a hypothesis. This is probably bad style so I'll try something else